File: backend/app/routers/score.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import text

from app.database import get_db
from app.schemas.schemas import ScoreCreate, ScoreResponse
from app.routers.auth import get_current_user

router = APIRouter(prefix="/api/score", tags=["成绩录入"])


@router.post("/entry", response_model=ScoreResponse)
async def entry_score(
    data: ScoreCreate,
    db: AsyncSession = Depends(get_db),
    current_user: dict = Depends(get_current_user),
):
    """录入高考成绩"""
    username = current_user["username"]
    await db.execute(
        text("INSERT INTO score (user_id, score, `rank`, year) VALUES (:uid, :score, :r, :year)"),
        {"uid": username, "score": data.score, "r": data.rank, "year": data.year or 2025}
    )
    await db.commit()
    new_id = (await db.execute(text("SELECT LAST_INSERT_ID()"))).scalar()
    return ScoreResponse(id=new_id, score=data.score, rank=data.rank, year=data.year or 2025, created_at=None)


@router.get("/history", response_model=list[ScoreResponse])
async def score_history(
    db: AsyncSession = Depends(get_db),
    current_user: dict = Depends(get_current_user),
):
    """获取成绩记录列表"""
    result = await db.execute(
        text("SELECT id, user_id, score, COALESCE(`rank`, 0), year, created_at "
             "FROM score WHERE user_id = :uid ORDER BY year DESC"),
        {"uid": current_user["username"]}
    )
    return [ScoreResponse(id=r[0], score=r[2], rank=r[3] if r[3] else None, year=r[4] or 2025, created_at=r[5]) for r in result.fetchall()]


@router.get("/latest", response_model=ScoreResponse)
async def score_latest(
    db: AsyncSession = Depends(get_db),
    current_user: dict = Depends(get_current_user),
):
    """获取最新成绩记录"""
    result = await db.execute(
        text("SELECT id, user_id, score, COALESCE(`rank`, 0), year, created_at "
             "FROM score WHERE user_id = :uid ORDER BY year DESC LIMIT 1"),
        {"uid": current_user["username"]}
    )
    row = result.fetchone()
    if not row:
        raise HTTPException(status_code=404, detail="暂无成绩记录")
    return ScoreResponse(id=row[0], score=row[2], rank=row[3] if row[3] else None, year=row[4] or 2025, created_at=row[5])

File: worker/src/routers/score.ts
import { Hono } from 'hono'
import type { Bindings } from '../types'
import { requireAuth } from '../middleware'

const app = new Hono<{ Bindings: Bindings }>()

async function listScores(db: D1Database, username: string, year?: string) {
  let sql = 'SELECT id, score, "rank", year, created_at FROM score WHERE user_id = ?'
  const params: any[] = [username]
  if (year) { sql += ' AND year = ?'; params.push(+year) }
  sql += ' ORDER BY year DESC'
  return (await db.prepare(sql).bind(...params).all()).results
}

// 前端: GET /api/score/history
app.get('/api/score/history', requireAuth, async (c) => {
  const results = await listScores(c.env.DB, c.var.auth.username, c.req.query('year'))
  return c.json(results)
})

// 前端: GET /api/score/latest
app.get('/api/score/latest', requireAuth, async (c) => {
  const row: any = await c.env.DB.prepare(
    'SELECT id, score, "rank", year, created_at FROM score WHERE user_id = ? ORDER BY year DESC LIMIT 1'
  ).bind(c.var.auth.username).first()
  if (!row) return c.json({ detail: '暂无成绩数据' }, 404)
  return c.json({ id: row.id, score: row.score, rank: row.rank, year: row.year, created_at: row.created_at })
})

// 前端: POST /api/score/entry  (与 /api/score/add 相同)
app.post('/api/score/entry', requireAuth, async (c) => {
  const { score, rank, year } = await c.req.json()
  await c.env.DB.prepare(
    'INSERT INTO score (user_id, score, "rank", year) VALUES (?, ?, ?, ?)'
  ).bind(c.var.auth.username, score, rank || null, year).run()
  return c.json({ message: '成绩录入成功' })
})

// 内部: GET /api/score/list
app.get('/api/score/list', requireAuth, async (c) => {
  const results = await listScores(c.env.DB, c.var.auth.username, c.req.query('year'))
  return c.json(results)
})

// 内部: POST /api/score/add
app.post('/api/score/add', requireAuth, async (c) => {
  const { score, rank, year } = await c.req.json()
  await c.env.DB.prepare(
    'INSERT INTO score (user_id, score, "rank", year) VALUES (?, ?, ?, ?)'
  ).bind(c.var.auth.username, score, rank, year).run()
  return c.json({ message: '成绩录入成功' })
})

app.put('/api/score/:score_id', requireAuth, async (c) => {
  const { score, rank, year } = await c.req.json()
  await c.env.DB.prepare(
    'UPDATE score SET score = ?, "rank" = ?, year = ? WHERE id = ? AND user_id = ?'
  ).bind(score, rank, year, +c.req.param('score_id'), c.var.auth.username).run()
  return c.json({ message: '成绩更新成功' })
})

app.delete('/api/score/:score_id', requireAuth, async (c) => {
  await c.env.DB.prepare(
    'DELETE FROM score WHERE id = ? AND user_id = ?'
  ).bind(+c.req.param('score_id'), c.var.auth.username).run()
  return c.json({ message: '成绩删除成功' })
})

export { app as scoreRouter }
